[wip] Add support for sbt files

jvican opened the pull request.

jvican committed:
Add scripted infrastructure to sbthost

jvican committed:
Test semanticdb for sbt files

jvican committed:
Handle dialects better

olafurpg commented on sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala:
```diff
@@ -74,11 +74,14 @@ trait SbthostPipeline extends DatabaseOps { self: SbthostPlugin =>
         }
+        // Select Sbt0137 dialect for scala sources extracted from sbt files
+        val detectedDialect =
+          if (g.getClass.getName.contains("sbt.compiler.Eval")) "Sbt0137" else "Scala210"
```
Nice!

olafurpg commented on sbthost/tests/src/test/scala/scala/meta/tests/SbthostTest.scala:
```diff
@@ -4,64 +4,128 @@ package tests
+         |Sbt0137
+         |
+         |Names:
+         |[0..12): organization => _root_._empty_.
```
Interesting, when running under sbt each statement becomes a separate compilation unit

```
[info] ====================
[info] = build.sbt
[info] ====================
[info] val p1 = project
[info]
[info]
[info]
[info] ====================
[info] = /private/var/folders/30/6jlz_xfs46ndvn212mt_wj6m0000gn/T/sbt_8a2705d2/basic/build.sbt
[info] ====================
[info] name := "basic"
[info]
[info]
[info] ====================
[info] = /private/var/folders/30/6jlz_xfs46ndvn212mt_wj6m0000gn/T/sbt_8a2705d2/basic/build.sbt
[info] ====================
[info] organization := "me.vican.jorge"
[info]
```

olafurpg replied:
We'll have to figure a way merge these into a single .semanticdb.

olafurpg replied:
Let's leave that for another PR.

jvican replied:
@olafurpg Yes, sbt creates a compilation unit for every unit. I have a WIP PR from two months ago that put everything into only one compilation unit, but it's not ready yet, so I think working around this issue in sbthost is the way to go :smile

olafurpg commented on sbthost/tests/src/test/scala/scala/meta/tests/SbthostTest.scala:
```diff
@@ -4,64 +4,128 @@ package tests
+         |[0..12): organization => _root_._empty_.
+         |[13..15): := => _root_.sbt.internals.DslEntry.fromSettingsDef(Lsbt/Init/SettingsDefinition;)Lsbt/internals/DslEntry;.
+         |
+         |Denotations:
```
I propose we leave denotations out of these tests

```scala
    val noDenotations = mirror.database.copy(entries = mirror.entries.map {
      case (input, attrs) =>
        input -> attrs.copy(denotations = Nil)
    })
    val syntax = noDenotations.syntax
    assert(syntax.trim == expectedSyntax.trim)
```

That avoid the need for the StableHashId and pre-process tricks, which was causing flakiness when I ran the tests multiple times locally. We are mostly interested in the positions in the Names section, anyways. WDYT?

jvican replied:
I thought about this too. I'm fine with this, I hope there are no future regressions!

olafurpg replied:
We can test the denotations separately for .scala files. For .sbt files the most important bit is the positions of names.

olafurpg reviewed (commented):
This is really nice, I can run `~test`, edit sbthost/nsc and "immediately" see the semanticdb from build.sbt. Awesome work with the sbt foo.

olafurpg committed:
Remove denotations from SbthostTest.

The positions in the names section is what we're most interested in. We
can test the denotations separately on .scala files, assuming they
will work similarly for .sbt files.

olafurpg committed:
Remove build.properties

olafurpg commented:
Superseded by #8
